### PR TITLE
echo: handle multibyte escape sequences

### PR DESCRIPTION
**Pull request by andrewliebenow**
Bug was reported, with root cause analysis, by kkew3 Added tests were derived from test cases provided by kkew3 See https://github.com/uutils/coreutils/issues/6741

### CODE REVIEW

File: src/uu/echo/src/echo.rs [hunk 1]
```diff
@@ -39,77 +47,117 @@ impl Base {
...
-    let mut ret = input.peek().and_then(|c| c.to_digit(base as u32))? as u8;
+    let radix = base.radix();
+    let radix_u_three_two = u32::from(radix);
+
+    let mut sum = match input.peek() {
+        Some(&&ue) => match char::from(ue).to_digit(radix_u_three_two) {
+            // A u8 interpreted as a hexadecimal or octal digit is never more than 16
+            Some(ut) => u8::try_from(ut).unwrap(),
+            None => {
+                return None;
```
Comment (sylvestre) [inline src/uu/echo/src/echo.rs:62]: please document this

File: src/uu/echo/src/echo.rs [hunk 2]
```diff
@@ -39,77 +47,117 @@ impl Base {
...
+            }
+        },
+        None => {
```
Comment (sylvestre) [inline src/uu/echo/src/echo.rs:65]: and this too

File: tests/by-util/test_echo.rs [hunk 3]
```diff
@@ -303,3 +303,86 @@ fn partial_version_argument() {
...
+    new_ucmd!()
+        .args(&["-e", r"\x41\xf0\x9f\x98\x82\x42"])
+        .succeeds()
+        .stdout_only("A\u{1F602}B\n");
+
+    new_ucmd!()
+        .args(&["-e", r"\xf0\x41\x9f\x98\x82"])
+        .succeeds()
+        .stdout_is_bytes(b"\xF0A\x9F\x98\x82\n")
+        .no_stderr();
```
Comment (cakebaker) [inline tests/by-util/test_echo.rs:333]: You can simplify this by using `stdout_only_bytes`.

File: src/uu/echo/src/echo.rs [hunk 4]
```diff
@@ -39,77 +47,117 @@ impl Base {
...
+fn parse_code(input: &mut Peekable<Iter<u8>>, base: Base) -> Option<u8> {
+    // All arithmetic on `sum` needs to be wrapping, because octal input can
     // take 3 digits, which is 9 bits, and therefore more than what fits in a
     // `u8`. GNU just seems to wrap these values.
-    // Note that if we instead make `ret` a `u32` and use `char::from_u32` will
-    // yield incorrect results because it will interpret values larger than
-    // `u8::MAX` as unicode.
-    let mut ret = input.peek().and_then(|c| c.to_digit(base as u32))? as u8;
+    let radix = base.radix();
+    let radix_u_three_two = u32::from(radix);
```
Comment (cakebaker) [inline src/uu/echo/src/echo.rs:55]: Writing `u32` as `u_three_two` looks strange ;-) Why not use `radix_u32`?

File: src/uu/echo/src/echo.rs [hunk 5]
```diff
@@ -39,77 +47,117 @@ impl Base {
...
+
+    let mut sum = match input.peek() {
+        Some(&&ue) => match char::from(ue).to_digit(radix_u_three_two) {
+            // A u8 interpreted as a hexadecimal or octal digit is never more than 16
+            Some(ut) => u8::try_from(ut).unwrap(),
```
Comment (cakebaker) [inline src/uu/echo/src/echo.rs:60]: What do `ue` and `ut` mean?

File: src/uu/echo/src/echo.rs [hunk 6]
```diff
@@ -39,77 +47,117 @@ impl Base {
...
-                'n' => '\n',
-                'r' => '\r',
-                't' => '\t',
-                'v' => '\x0b',
-                'x' => {
-                    if let Some(c) = parse_code(&mut iter, Base::Hex) {
-                        c
+            // For extending lifetime
+            let sl: [u8; 1_usize];
+            let sli: [u8; 2_usize];
```
Comment (cakebaker) [inline src/uu/echo/src/echo.rs:120]: Why are those variables needed? Plus the variable names are not that great ;-)

File: src/uu/echo/src/echo.rs [hunk 7]
```diff
@@ -39,77 +47,117 @@ impl Base {
...
+                b'b' => b"\x08",
+                b'c' => return Ok(ControlFlow::Break(())),
+                b'e' => b"\x1b",
+                b'f' => b"\x0c",
+                b'n' => b"\n",
+                b'r' => b"\r",
+                b't' => b"\t",
+                b'v' => b"\x0b",
+                b'x' => {
+                    if let Some(ue) = parse_code(&mut iter, Base::Hex) {
```
Comment (cakebaker) [inline src/uu/echo/src/echo.rs:134]: What does `ue` mean?

File: src/uu/echo/src/echo.rs [hunk 8]
```diff
@@ -120,15 +168,28 @@ fn print_escaped(input: &str, mut output: impl Write) -> io::Result<ControlFlow<
...
-        Some(s) => s.map(|s| s.to_string()).collect(),
-        None => vec![String::new()],
-    };
 
-    execute(no_newline, escaped, &values)
-        .map_err_context(|| "could not write to stdout".to_string())
+    let mut stdout_lock = io::stdout().lock();
+
+    match matches.get_many::<OsString>(options::STRING) {
+        Some(va) => {
```
Comment (cakebaker) [inline src/uu/echo/src/echo.rs:181]: What does `va` mean?

File: src/uu/echo/src/echo.rs (hunk 6 above)
Comment (andrewliebenow) [reply src/uu/echo/src/echo.rs:120]: Will fix the names. It's an MSRV issue. Until Rust 1.79, temporary lifetime extension didn't work in if and match:

https://github.com/rust-lang/rust/pull/121346
https://releases.rs/docs/1.79.0/

Comment (cakebaker) [reply src/uu/echo/src/echo.rs:120]: Ah, ok. Can you please add a todo to remove those vars in the future when we have an MSRV of 1.79?

**Review by cakebaker — APPROVED** [review]

**Comment by cakebaker** [comment]
Thanks!

**Pull request merged**